File: src/main/features/parser/LinkParser.ts
import { ParsedLink, Platform, ContentType } from '../../../shared/types'
import crypto from 'crypto'

export class LinkParser {
  private static readonly YOUTUBE_VIDEO_REGEX = /https?:\/\/(?:www\.)?youtube\.com\/watch\?(?:.*?[&])?v=([a-zA-Z0-9_-]{11})/i
  private static readonly YOUTUBE_SHORT_REGEX = /https?:\/\/(?:www\.)?youtube\.com\/shorts\/([a-zA-Z0-9_-]{11})/i
  private static readonly YOUTU_BE_REGEX = /https?:\/\/youtu\.be\/([a-zA-Z0-9_-]{11})/i
  private static readonly YOUTUBE_PLAYLIST_REGEX = /https?:\/\/(?:www\.)?youtube\.com\/playlist\?(?:.*?[&])?list=([a-zA-Z0-9_-]+)/i
  private static readonly YOUTUBE_MUSIC_REGEX = /https?:\/\/music\.youtube\.com\/watch\?(?:.*?[&])?v=([a-zA-Z0-9_-]{11})/i
  private static readonly YOUTUBE_MUSIC_PLAYLIST_REGEX = /https?:\/\/music\.youtube\.com\/playlist\?(?:.*?[&])?list=([a-zA-Z0-9_-]+)/i
  private static readonly YOUTUBE_MUSIC_ALBUM_REGEX = /https?:\/\/music\.youtube\.com\/(?:album|browse)\/([a-zA-Z0-9_-]+)/i
  private static readonly YOUTUBE_CHANNEL_REGEX = /https?:\/\/(?:music\.|www\.)?youtube\.com\/(?:@|channel\/|c\/)([a-zA-Z0-9_.-]+)/i
  private static readonly PLAYLIST_PARAM_REGEX = /[?&]list=([a-zA-Z0-9_-]+)/i

  /**
   * Parses an input string (URL or search query) into a ParsedLink object.
   */
  public static parse(inputUrl: string): ParsedLink | null {
    const trimmed = (inputUrl || '').trim()
    if (!trimmed) return null

    // 1. YouTube / YouTube Music Playlist (Check query param first)
    const playlistMatch = trimmed.match(this.PLAYLIST_PARAM_REGEX)
    if (playlistMatch && playlistMatch[1]) {
      const isMusic = trimmed.includes('music.youtube.com')
      return {
        platform: isMusic ? 'youtube_music' : 'youtube',
        contentType: 'playlist',
        id: playlistMatch[1],
        originalUrl: trimmed
      }
    }

    // 2. YouTube Music Album / Playlist
    const ytMusicPlaylistMatch = trimmed.match(this.YOUTUBE_MUSIC_PLAYLIST_REGEX)
    if (ytMusicPlaylistMatch && ytMusicPlaylistMatch[1]) {
      return {
        platform: 'youtube_music',
        contentType: 'playlist',
        id: ytMusicPlaylistMatch[1],
        originalUrl: trimmed
      }
    }

    const ytMusicAlbumMatch = trimmed.match(this.YOUTUBE_MUSIC_ALBUM_REGEX)
    if (ytMusicAlbumMatch && ytMusicAlbumMatch[1]) {
      return {
        platform: 'youtube_music',
        contentType: 'album',
        id: ytMusicAlbumMatch[1],
        originalUrl: trimmed
      }
    }

    // 3. YouTube Music Track
    const ytMusicTrackMatch = trimmed.match(this.YOUTUBE_MUSIC_REGEX)
    if (ytMusicTrackMatch && ytMusicTrackMatch[1]) {
      return {
        platform: 'youtube_music',
        contentType: 'track',
        id: ytMusicTrackMatch[1],
        originalUrl: trimmed
      }
    }

    // 4. Standard YouTube Playlist
    const ytPlaylistMatch = trimmed.match(this.YOUTUBE_PLAYLIST_REGEX)
    if (ytPlaylistMatch && ytPlaylistMatch[1]) {
      return {
        platform: 'youtube',
        contentType: 'playlist',
        id: ytPlaylistMatch[1],
        originalUrl: trimmed
      }
    }

    // 5. YouTube Shorts / Watch / youtu.be
    const ytShortMatch = trimmed.match(this.YOUTUBE_SHORT_REGEX)
    if (ytShortMatch && ytShortMatch[1]) {
      return {
        platform: 'youtube',
        contentType: 'video',
        id: ytShortMatch[1],
        originalUrl: trimmed
      }
    }

    const youtuBeMatch = trimmed.match(this.YOUTU_BE_REGEX)
    if (youtuBeMatch && youtuBeMatch[1]) {
      return {
        platform: 'youtube',
        contentType: 'video',
        id: youtuBeMatch[1],
        originalUrl: trimmed
      }
    }

    const ytVideoMatch = trimmed.match(this.YOUTUBE_VIDEO_REGEX)
    if (ytVideoMatch && ytVideoMatch[1]) {
      return {
        platform: 'youtube',
        contentType: 'video',
        id: ytVideoMatch[1],
        originalUrl: trimmed
      }
    }

    // 6. YouTube Channel / Artist
    const ytChannelMatch = trimmed.match(this.YOUTUBE_CHANNEL_REGEX)
    if (ytChannelMatch && ytChannelMatch[1]) {
      const isMusic = trimmed.includes('music.youtube.com')
      return {
        platform: isMusic ? 'youtube_music' : 'youtube',
        contentType: 'playlist',
        id: ytChannelMatch[1],
        originalUrl: trimmed
      }
    }

    // 7. Generic URL fallback
    if (trimmed.startsWith('http://') || trimmed.startsWith('https://')) {
      const hashId = crypto.createHash('md5').update(trimmed).digest('hex').substring(0, 12)
      return {
        platform: 'other',
        contentType: 'video',
        id: hashId,
        originalUrl: trimmed
      }
    }

    // 8. Plain Text Artist / Query search fallback
    return {
      platform: 'youtube',
      contentType: 'playlist',
      id: trimmed,
      originalUrl: `ytsearch150:${trimmed} music.youtube.com`
    }
  }

  public static isSupported(url: string): boolean {
    return this.parse(url) !== null
  }
}

File: Stash-Android/app/src/main/java/com/eurtlabs/stash/data/parser/LinkParser.kt
package com.eurtlabs.stash.data.parser

import com.eurtlabs.stash.data.model.ContentType
import com.eurtlabs.stash.data.model.ParsedLink
import com.eurtlabs.stash.data.model.Platform
import java.security.MessageDigest
import java.util.regex.Pattern

object LinkParser {

    private val YOUTUBE_VIDEO_REGEX = Pattern.compile("https?://(?:www\\.)?youtube\\.com/watch\\?(?:.*?[&])?v=([a-zA-Z0-9_-]{11})", Pattern.CASE_INSENSITIVE)
    private val YOUTUBE_SHORT_REGEX = Pattern.compile("https?://(?:www\\.)?youtube\\.com/shorts/([a-zA-Z0-9_-]{11})", Pattern.CASE_INSENSITIVE)
    private val YOUTU_BE_REGEX = Pattern.compile("https?://youtu\\.be/([a-zA-Z0-9_-]{11})", Pattern.CASE_INSENSITIVE)
    private val YOUTUBE_PLAYLIST_REGEX = Pattern.compile("https?://(?:www\\.)?youtube\\.com/playlist\\?(?:.*?[&])?list=([a-zA-Z0-9_-]+)", Pattern.CASE_INSENSITIVE)
    private val YOUTUBE_MUSIC_REGEX = Pattern.compile("https?://music\\.youtube\\.com/watch\\?(?:.*?[&])?v=([a-zA-Z0-9_-]{11})", Pattern.CASE_INSENSITIVE)
    private val YOUTUBE_MUSIC_PLAYLIST_REGEX = Pattern.compile("https?://music\\.youtube\\.com/playlist\\?(?:.*?[&])?list=([a-zA-Z0-9_-]+)", Pattern.CASE_INSENSITIVE)
    private val YOUTUBE_MUSIC_ALBUM_REGEX = Pattern.compile("https?://music\\.youtube\\.com/(?:album|browse)/([a-zA-Z0-9_-]+)", Pattern.CASE_INSENSITIVE)
    private val YOUTUBE_CHANNEL_REGEX = Pattern.compile("https?://(?:music\\.|www\\.)?youtube\\.com/(?:@|channel/|c/)([a-zA-Z0-9_.-]+)", Pattern.CASE_INSENSITIVE)
    private val PLAYLIST_PARAM_REGEX = Pattern.compile("[?&]list=([a-zA-Z0-9_-]+)", Pattern.CASE_INSENSITIVE)

    fun parse(inputUrl: String?): ParsedLink? {
        val trimmed = inputUrl?.trim() ?: return null
        if (trimmed.isEmpty()) return null

        // 1. YouTube / YouTube Music Playlist (Query param check)
        val playlistMatcher = PLAYLIST_PARAM_REGEX.matcher(trimmed)
        if (playlistMatcher.find()) {
            val isMusic = trimmed.contains("music.youtube.com")
            return ParsedLink(
                platform = if (isMusic) Platform.YOUTUBE_MUSIC else Platform.YOUTUBE,
                contentType = ContentType.PLAYLIST,
                id = playlistMatcher.group(1) ?: "",
                originalUrl = trimmed
            )
        }

        // 2. YouTube Music Album / Playlist
        val ytMusicPlaylistMatcher = YOUTUBE_MUSIC_PLAYLIST_REGEX.matcher(trimmed)
        if (ytMusicPlaylistMatcher.find()) {
            return ParsedLink(
                platform = Platform.YOUTUBE_MUSIC,
                contentType = ContentType.PLAYLIST,
                id = ytMusicPlaylistMatcher.group(1) ?: "",
                originalUrl = trimmed
            )
        }

        val ytMusicAlbumMatcher = YOUTUBE_MUSIC_ALBUM_REGEX.matcher(trimmed)
        if (ytMusicAlbumMatcher.find()) {
            return ParsedLink(
                platform = Platform.YOUTUBE_MUSIC,
                contentType = ContentType.ALBUM,
                id = ytMusicAlbumMatcher.group(1) ?: "",
                originalUrl = trimmed
            )
        }

        // 3. YouTube Music Track
        val ytMusicTrackMatcher = YOUTUBE_MUSIC_REGEX.matcher(trimmed)
        if (ytMusicTrackMatcher.find()) {
            return ParsedLink(
                platform = Platform.YOUTUBE_MUSIC,
                contentType = ContentType.TRACK,
                id = ytMusicTrackMatcher.group(1) ?: "",
                originalUrl = trimmed
            )
        }

        // 4. Standard YouTube Playlist
        val ytPlaylistMatcher = YOUTUBE_PLAYLIST_REGEX.matcher(trimmed)
        if (ytPlaylistMatcher.find()) {
            return ParsedLink(
                platform = Platform.YOUTUBE,
                contentType = ContentType.PLAYLIST,
                id = ytPlaylistMatcher.group(1) ?: "",
                originalUrl = trimmed
            )
        }

        // 5. YouTube Shorts / Watch / youtu.be
        val ytShortMatcher = YOUTUBE_SHORT_REGEX.matcher(trimmed)
        if (ytShortMatcher.find()) {
            return ParsedLink(
                platform = Platform.YOUTUBE,
                contentType = ContentType.VIDEO,
                id = ytShortMatcher.group(1) ?: "",
                originalUrl = trimmed
            )
        }

        val youtuBeMatcher = YOUTU_BE_REGEX.matcher(trimmed)
        if (youtuBeMatcher.find()) {
            return ParsedLink(
                platform = Platform.YOUTUBE,
                contentType = ContentType.VIDEO,
                id = youtuBeMatcher.group(1) ?: "",
                originalUrl = trimmed
            )
        }

        val ytVideoMatcher = YOUTUBE_VIDEO_REGEX.matcher(trimmed)
        if (ytVideoMatcher.find()) {
            return ParsedLink(
                platform = Platform.YOUTUBE,
                contentType = ContentType.VIDEO,
                id = ytVideoMatcher.group(1) ?: "",
                originalUrl = trimmed
            )
        }

        // 6. YouTube Channel / Artist
        val ytChannelMatcher = YOUTUBE_CHANNEL_REGEX.matcher(trimmed)
        if (ytChannelMatcher.find()) {
            val isMusic = trimmed.contains("music.youtube.com")
            return ParsedLink(
                platform = if (isMusic) Platform.YOUTUBE_MUSIC else Platform.YOUTUBE,
                contentType = ContentType.PLAYLIST,
                id = ytChannelMatcher.group(1) ?: "",
                originalUrl = trimmed
            )
        }

        // 7. Generic URL fallback (only valid http/https URLs)
        if (trimmed.startsWith("http://", ignoreCase = true) || trimmed.startsWith("https://", ignoreCase = true)) {
            // Validate it looks like an actual URL (has a domain)
            val domainMatch = Regex("^https?://[a-zA-Z0-9]").containsMatchIn(trimmed)
            if (domainMatch) {
                val hashId = md5(trimmed).take(12)
                return ParsedLink(
                    platform = Platform.OTHER,
                    contentType = ContentType.VIDEO,
                    id = hashId,
                    originalUrl = trimmed
                )
            }
        }

        // 8. Reject file paths, random non-URL text
        // Only allow input that doesn't contain backslashes or look like a local path
        if (trimmed.contains("\\") || trimmed.contains(":/") || trimmed.startsWith("/") || trimmed.startsWith("C:")) {
            return null
        }

        // 9. Plain text query — search fallback (only for clean query strings)
        if (trimmed.length in 2..120 && !trimmed.contains("://")) {
            return ParsedLink(
                platform = Platform.YOUTUBE,
                contentType = ContentType.TRACK,
                id = trimmed,
                originalUrl = "ytsearch1:$trimmed"
            )
        }

        return null
    }

    private fun md5(input: String): String {
        val md = MessageDigest.getInstance("MD5")
        return md.digest(input.toByteArray()).joinToString("") { "%02x".format(it) }
    }
}
